torchvision/prototype/datasets/_builtin/oxford_iiit_pet.py

```diff
@@ -0,0 +1,150 @@
+            species="cat" if data["species"] == "1" else "dog",
+        )
+
+    def _collate_and_decode_sample(
+        self,
+        data: Tuple[Tuple[Dict[str, str], Tuple[str, io.IOBase]], Tuple[str, io.IOBase]],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> Dict[str, Any]:
+        ann_data, image_data = data
+        classification_data, segmentation_data = ann_data
+        segmentation_path, segmentation_buffer = segmentation_data
+        image_path, image_buffer = image_data
+
+        return dict(
+            self._decode_classification_data(classification_data),
+            segmentation_path=segmentation_path,
+            segmentation=decoder(segmentation_buffer) if decoder else segmentation_buffer,
+            image_path=image_path,
+            image=decoder(image_buffer) if decoder else image_buffer,
+        )
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Dict[str, Any]]:
+        images_dp, anns_dp = resource_dps
+
+        images_dp = Filter(images_dp, self._filter_images)
+
+        split_and_classification_dp, segmentations_dp = Demultiplexer(
+            anns_dp,
+            2,
+            self._classify_anns,
+            drop_none=True,
+            buffer_size=INFINITE_BUFFER_SIZE,
+        )
+
+        split_and_classification_dp = Filter(
+            split_and_classification_dp, path_comparator("name", f"{config.split}.txt")
+        )
+        split_and_classification_dp = CSVDictParser(
+            split_and_classification_dp, fieldnames=("image_id", "label", "species"), delimiter=" "
+        )
+        split_and_classification_dp = hint_sharding(split_and_classification_dp)
+        split_and_classification_dp = hint_shuffling(split_and_classification_dp)
+
+        segmentations_dp = Filter(segmentations_dp, self._filter_segmentations)
+
+        anns_dp = IterKeyZipper(
+            split_and_classification_dp,
+            segmentations_dp,
+            key_fn=getitem("image_id"),
+            ref_key_fn=path_accessor("stem"),
+            buffer_size=INFINITE_BUFFER_SIZE,
+        )
+
+        dp = IterKeyZipper(
+            anns_dp,
+            images_dp,
+            key_fn=getitem(0, "image_id"),
+            ref_key_fn=path_accessor("stem"),
+            buffer_size=INFINITE_BUFFER_SIZE,
+        )
+        return Mapper(dp, functools.partial(self._collate_and_decode_sample, decoder=decoder))
+
+    def _filter_split_and_classification_anns(self, data: Tuple[str, Any]) -> bool:
+        return self._classify_anns(data) == 0
+
+    def _generate_categories(self, root: pathlib.Path) -> List[str]:
+        config = self.default_config
+        dp = self.resources(config)[1].load(pathlib.Path(root) / self.name)
+        dp = Filter(dp, self._filter_split_and_classification_anns)
```

Comment:
Unfortunately, datapipes do not allow lambdas or local functions at callables, because they cannot safely be serialized. Thus, we actually need either standalone functions or methods here. I opted for the latter since they are closely tied to the class.

For simple instances we have some helper methods like `getitem` or `path_accessor` (see L131-L132). For more "complex" cases we need to define it ourselves.